No unique bean of type RubyLocatorServiceFactory

Hi Wolf,

maybe this is an Bamboo v4.1 support issue or a misconfiguration with my setup. It's my first try to integrate a rails project with bamboo und I really appreciate your plugin and would like to get it to work. The Bamboo Agent does not start if your plugin is activated on the server.

Any hints are very welcome.

With kind regards, Ingo

```
INFO   | jvm 1    | 2012/06/12 12:04:31 | Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'au.id.wolfe.bamboo.ruby.RubyCapabilityDefaultsHelper': Unsatisfied dependency express
ed through constructor argument with index 0 of type [au.id.wolfe.bamboo.ruby.locator.RubyLocatorServiceFactory]: : No unique bean of type [au.id.wolfe.bamboo.ruby.locator.RubyLocatorServiceFactory] is defined: Unsatisfied dependency o
f type [class au.id.wolfe.bamboo.ruby.locator.RubyLocatorServiceFactory]: expected at least 1 matching bean; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No unique bean of type [au.id.wolfe.bambo
o.ruby.locator.RubyLocatorServiceFactory] is defined: Unsatisfied dependency of type [class au.id.wolfe.bamboo.ruby.locator.RubyLocatorServiceFactory]: expected at least 1 matching bean
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:591)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:193)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:925)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:835)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:440)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory$1.run(AbstractAutowireCapableBeanFactory.java:409)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at java.security.AccessController.doPrivileged(Native Method)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:380)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:300)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)INFO   | jvm 1    | 2012/06/12 12:04:31 |       at java.lang.reflect.Method.invoke(Method.java:616)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at com.atlassian.plugin.osgi.spring.DefaultSpringContainerAccessor.createBean(DefaultSpringContainerAccessor.java:109)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at com.atlassian.bamboo.plugin.descriptor.AbstractBambooModuleDescriptor.instantiateClass(AbstractBambooModuleDescriptor.java:104)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at com.atlassian.bamboo.plugin.descriptor.TaskModuleDescriptorImpl$2.create(TaskModuleDescriptorImpl.java:69)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at com.atlassian.bamboo.plugin.descriptor.TaskModuleDescriptorImpl$2.create(TaskModuleDescriptorImpl.java:62)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at com.atlassian.util.concurrent.ResettableLazyReference$InternalReference.create(ResettableLazyReference.java:143)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at com.atlassian.util.concurrent.LazyReference$Sync.run(LazyReference.java:326)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at com.atlassian.util.concurrent.LazyReference.getInterruptibly(LazyReference.java:146)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       ... 30 more
INFO   | jvm 1    | 2012/06/12 12:04:31 | Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No unique bean of type [au.id.wolfe.bamboo.ruby.locator.RubyLocatorServiceFactory] is defined: Unsatisfied dependency
 of type [class au.id.wolfe.bamboo.ruby.locator.RubyLocatorServiceFactory]: expected at least 1 matching bean
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:613)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:622)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:584)
INFO   | jvm 1    | 2012/06/12 12:04:31 |       ... 49 more
ERROR  | wrapper  | 2012/06/12 12:04:32 | JVM exited unexpectedly.
```

--- a/pom.xml
+++ b/pom.xml
@@ -6,7 +6,7 @@
     <artifactId>rake-bamboo-plugin</artifactId> <!-- Leaving this the OLD name to ensure we don't loose anyone -->
     <version>1.8-SNAPSHOT</version>
     <name>Bamboo Ruby Plugin</name>
-    <description>Ruby plugin for Atlassian Bamboo.</description>
+    <description>Plugin for Atlassian Bamboo which contains an array of ruby tasks.</description>
     <packaging>atlassian-plugin</packaging>
 
     <organization>
@@ -37,7 +37,7 @@
     </scm>
 
     <properties>
-        <bamboo.version>4.0</bamboo.version>
+        <bamboo.version>4.1</bamboo.version>
         <bamboo.data.version>3.2.2</bamboo.data.version>
         <amps.version>3.7</amps.version>
     </properties>
@@ -77,7 +77,20 @@
             <version>4.6</version>
             <scope>test</scope>
         </dependency>
+
     </dependencies>
+
+    <dependencyManagement>
+        <dependencies>
+
+            <dependency>
+                <groupId>org.slf4j</groupId>
+                <artifactId>slf4j-log4j12</artifactId>
+                <version>1.6.4</version>
+            </dependency>
+
+        </dependencies>
+    </dependencyManagement>
 
     <build>
         <plugins>

--- a/src/main/java/au/id/wolfe/bamboo/ruby/RubyCapabilityDefaultsHelper.java
+++ b/src/main/java/au/id/wolfe/bamboo/ruby/RubyCapabilityDefaultsHelper.java
@@ -24,6 +24,10 @@ public class RubyCapabilityDefaultsHelper implements CapabilityDefaultsHelper {
 
     private final RubyLocatorServiceFactory rubyLocatorServiceFactory;
 
+    public RubyCapabilityDefaultsHelper() {
+        this.rubyLocatorServiceFactory = new RubyLocatorServiceFactory();
+    }
+
     public RubyCapabilityDefaultsHelper(RubyLocatorServiceFactory rubyLocatorServiceFactory) {
         log.info("loading rubyLocatorServiceFactory = " + rubyLocatorServiceFactory);
         this.rubyLocatorServiceFactory = rubyLocatorServiceFactory;

--- a/src/main/java/au/id/wolfe/bamboo/ruby/locator/RubyLocatorServiceFactory.java
+++ b/src/main/java/au/id/wolfe/bamboo/ruby/locator/RubyLocatorServiceFactory.java
@@ -20,6 +20,14 @@ public class RubyLocatorServiceFactory {
 
     private final List<RubyRuntimeLocatorService> rubyRuntimeLocatorServices;
 
+    public RubyLocatorServiceFactory() {
+        rubyRuntimeLocatorServices = Lists.newArrayList(new SystemRubyRuntimeLocatorService(),
+                new RvmRubyRuntimeLocatorService(),
+                new WindowsRubyRuntimeLocatorService());
+
+        log.info("Loaded ruby runtime managers {} ", rubyRuntimeLocatorServices);
+    }
+
     public RubyLocatorServiceFactory(SystemRubyRuntimeLocatorService systemRubyRuntimeLocatorService,
                                      RvmRubyRuntimeLocatorService rvmRubyRuntimeLocatorService,
                                      WindowsRubyRuntimeLocatorService windowsRubyRuntimeLocatorService) {
@@ -28,7 +36,7 @@ public class RubyLocatorServiceFactory {
                 rvmRubyRuntimeLocatorService,
                 windowsRubyRuntimeLocatorService);
 
-        log.info("Loading ruby runtime managers {} ", rubyRuntimeLocatorServices);
+        log.info("Loaded ruby runtime managers {} ", rubyRuntimeLocatorServices);
 
     }
 

--- a/src/main/java/au/id/wolfe/bamboo/ruby/system/SystemRubyRuntimeLocatorService.java
+++ b/src/main/java/au/id/wolfe/bamboo/ruby/system/SystemRubyRuntimeLocatorService.java
@@ -14,6 +14,10 @@ public class SystemRubyRuntimeLocatorService implements RubyRuntimeLocatorServic
 
     private final SystemRubyLocator systemRubyLocator;
 
+    public SystemRubyRuntimeLocatorService() {
+        systemRubyLocator = new SystemRubyLocator(new FileSystemHelper());
+    }
+
     public SystemRubyRuntimeLocatorService(FileSystemHelper fileSystemHelper) {
         systemRubyLocator = new SystemRubyLocator(fileSystemHelper);
     }

--- a/src/main/java/au/id/wolfe/bamboo/ruby/tasks/AbstractRubyTask.java
+++ b/src/main/java/au/id/wolfe/bamboo/ruby/tasks/AbstractRubyTask.java
@@ -85,6 +85,9 @@ public abstract class AbstractRubyTask implements TaskType {
     protected abstract List<String> buildCommandList(RubyLabel rubyRuntimeLabel, ConfigurationMap config);
 
     protected RubyLocator getRubyLocator(String rubyRuntimeManager) {
+        if (rubyLocatorServiceFactory == null) {
+            rubyLocatorServiceFactory = new RubyLocatorServiceFactory();
+        }
         return rubyLocatorServiceFactory.acquireRubyLocator(rubyRuntimeManager);
     }
 

--- a/src/main/java/au/id/wolfe/bamboo/ruby/windows/WindowsRubyRuntimeLocatorService.java
+++ b/src/main/java/au/id/wolfe/bamboo/ruby/windows/WindowsRubyRuntimeLocatorService.java
@@ -15,6 +15,10 @@ public class WindowsRubyRuntimeLocatorService implements RubyRuntimeLocatorServi
 
     private final WindowsRubyLocator windowsRubyLocator;
 
+    public WindowsRubyRuntimeLocatorService() {
+        this.windowsRubyLocator = new WindowsRubyLocator(new ExecHelper());
+    }
+
     public WindowsRubyRuntimeLocatorService(ExecHelper execHelper) {
         this.windowsRubyLocator = new WindowsRubyLocator(execHelper);
     }

--- a/src/main/resources/atlassian-plugin.xml
+++ b/src/main/resources/atlassian-plugin.xml
@@ -5,24 +5,6 @@
         <vendor name="${project.organization.name}" url="${project.organization.url}"/>
     </plugin-info>
 
-    <component key="execHelper" name="Exec Helper"
-               class="au.id.wolfe.bamboo.ruby.util.ExecHelper"/>
-
-    <component key="fileSystemHelper" name="File System Helper"
-               class="au.id.wolfe.bamboo.ruby.util.FileSystemHelper"/>
-
-    <component key="rvmLocatorService" name="RVM Locator Service"
-        class="au.id.wolfe.bamboo.ruby.rvm.RvmRubyRuntimeLocatorService" />
-
-    <component key="windowsLocatorService" name="Windows Ruby Locator Service"
-               class="au.id.wolfe.bamboo.ruby.windows.WindowsRubyRuntimeLocatorService" />
-
-    <component key="systemLocatorService" name="System Ruby Locator Service"
-               class="au.id.wolfe.bamboo.ruby.system.SystemRubyRuntimeLocatorService" />
-
-    <component key="rubyLocatorServiceFactory" name="Ruby Locator Service Factory"
-               class="au.id.wolfe.bamboo.ruby.locator.RubyLocatorServiceFactory" />
-
     <taskType name="Bundler" class="au.id.wolfe.bamboo.ruby.tasks.bundler.BundlerTask" key="task.builder.bundler">
 
         <description>Use Bundler to install project dependencies.</description>

--- a/src/main/resources/bundler/editBundlerTask.ftl
+++ b/src/main/resources/bundler/editBundlerTask.ftl
@@ -4,8 +4,8 @@
             list=uiConfigBean.getExecutableLabels('ruby')
             extraUtility=addExecutableLink required='true' /]
 
-[@ww.textfield labelKey='bundle.path' name='path' required='false' cssClass="long-field" /]
+[@ww.textfield labelKey='bundler.path' name='path' required='false' cssClass="long-field" /]
 
-[@ww.checkbox labelKey='bundle.binstubs'
+[@ww.checkbox labelKey='bundler.binstubs'
                   name='binstubs'
                   toggle='true' /]

--- a/src/main/resources/english.properties
+++ b/src/main/resources/english.properties
@@ -13,7 +13,7 @@ rake.verbose.description=Enable verbose output.
 bundler.binstubs=Bin Stubs
 bundler.binstubs.description = Create a directory (defaults to bin) containing  an  executable that  runs  in  the  context of the bundle.
 bundler.path=Path
-bundler.path.description = The location to install the gems in the bundle to.
+bundler.path.description = Install the gems in the bundle to this path.
 bundler.ruby.description = Ruby environment to use for bundler.
 
 ruby.runtime=Ruby Environment

--- a/src/test/java/au/id/wolfe/bamboo/ruby/system/SystemRubyLocatorTest.java
+++ b/src/test/java/au/id/wolfe/bamboo/ruby/system/SystemRubyLocatorTest.java
@@ -37,7 +37,7 @@ public class SystemRubyLocatorTest {
 
         assertThat(rubyRuntimeList.size(), equalTo(1));
 
-        assertThat(rubyRuntimeList.get(0).getRubyName(), equalTo("1.8.7-p357"));
+        assertThat(rubyRuntimeList.get(0).getRubyName(), equalTo("1.8.7-p352"));
 
 
     }

